Frameworks/UIKit/UIView.mm

```diff
@@ -274,30 +260,106 @@ - (bool)_processGesturesForTouch:(UITouch*)touch event:(UIEvent*)event touchEven
         }
     }
 
-    // gesture priority list
-    const static id s_gesturesPriority[] = {[UIPinchGestureRecognizer class], [UISwipeGestureRecognizer class],
-                                            [UIPanGestureRecognizer class],   [UILongPressGestureRecognizer class],
-                                            [UITapGestureRecognizer class],   [_UIDMPanGestureRecognizer class] };
+    //
+    // After state transtion, we potentially get active gestures, we should remove any gestures that can not be recognized simultaneously
+    //
+    NSMutableSet* gesturesToRemove = [NSMutableSet set];
+    for (UIGestureRecognizer* curGesture in [g_currentlyTrackingGesturesList reverseObjectEnumerator]) {
+        UIGestureRecognizerState state = curGesture.state;
+        if (state != UIGestureRecognizerStatePossible && state != UIGestureRecognizerStateCancelled) {
+            UIGestureRecognizer* activeGesture = curGesture;
 
-    const static int s_numGestureTypes = sizeof(s_gesturesPriority) / sizeof(s_gesturesPriority[0]);
+            BOOL activeResponds = [activeGesture.delegate
+                respondsToSelector:@selector(gestureRecognizer:shouldRecognizeSimultaneouslyWithGestureRecognizer:)];
 
-    //  Process all gestures, including DM gesture
-    for (auto const& curgestureClass : s_gesturesPriority) {
-        id gestures = [g_curGesturesDict objectForKey:curgestureClass];
-        if ([curgestureClass _fireGestures:gestures shouldCancelTouches:shouldCancelTouches]) {
-            if (DEBUG_GESTURES) {
+            // now test it against the other gestures, removing if we do not handle simultaneous gestures
+            for (UIGestureRecognizer* curGesture in g_currentlyTrackingGesturesList) {
+                if (curGesture == activeGesture) {
+                    continue;
+                }
+
+                if (curGesture.state != UIGestureRecognizerStatePossible && state != UIGestureRecognizerStateCancelled) {
+                    // curGesture already active
+                    continue;
+                }
+
+                // Based on reference platform: test if active gesture can prevent other gesture, if so, further check if the other gesture
+                // can be prevented by active
+                // gesture if both result are YES, we need consult the delegates for both gesture. if both delegates saying NO for
+                // concurrent firing
+                // the other gesture will be prevented from firing by active gesture.
+                if ([activeGesture canPreventGestureRecognizer:curGesture] &&
+                    [curGesture canBePreventedByGestureRecognizer:activeGesture]) {
+                    BOOL currentResponds = [curGesture.delegate
+                        respondsToSelector:@selector(gestureRecognizer:shouldRecognizeSimultaneouslyWithGestureRecognizer:)];
+                    if ((currentResponds == NO ||
```

Comment:
>currentResponds == NO [](start = 25, length = 21)

really? If they *don't* respond we default to not simultaneous?  is that the expected behavior? if so, that's probably worth calling out as a comment